src/coreclr/src/vm/nativeimage.h

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _NATIVEIMAGE_H
+#define _NATIVEIMAGE_H
+
+#include "readytoruninfo.h"
+
+/// <summary>
+/// This structure is used in NativeImage to map simple names of component assemblies
+/// to their indices within the component assembly header table.
+/// </summary>
+struct AssemblyNameIndex
+{
+    SString Name;
+    int32_t Index;
+    
+    AssemblyNameIndex() : Index(-1) {}
+    AssemblyNameIndex(const SString& name, int32_t index);
+};
+
+class AssemblyNameIndexHashTraits : public NoRemoveSHashTraits< PtrSHashWithCleanupTraits<AssemblyNameIndex, const SString&> >
+{
+public:
+    static AssemblyNameIndex *Null() { return NULL; }
+    static bool IsNull(const AssemblyNameIndex *e) { return e == NULL; }
+
+    static const SString& GetKey(const AssemblyNameIndex *assemblyNameIndex) { return assemblyNameIndex->Name; }
+    static BOOL Equals(const SString& a, const SString& b) { return a.Equals(b); }
+    static count_t Hash(const SString& a) { return a.Hash(); }
+};
+
+class AssemblyLoadContext;
+class ReadyToRunInfo;
+class PEFile;
+class PEImage;
+
+/// <summary>
+/// This class represents a native ReadyToRun image. As of today, this file format is used
+/// as the compiled native code cache in composite R2R Crossgen2 build mode. Moving forward
+/// we plan to add support for OS-specific native executables (ELF on Linux, MachO on OSX).
+///
+/// The native image is identified by a well-known public export 'RTR_HEADER' pointing to the
+/// master READYTORUN_HEADER structure for the entire file. For composite R2R executables
+/// built by crossgenning a larger number of input MSIL assemblies the READYTORUN_HEADER
+/// contains a section named ComponentAssemblies that points to READYTORUN_CORE_HEADER
+/// structures representing the individual component assemblies and per-assembly sections.
+/// </summary>
+class NativeImage
+{
+private:
+    AssemblyLoadContext *m_pLoadContext;
+    LPCUTF8 m_utf8SimpleName;
+    bool m_runEagerFixups;
+    
+    NewHolder<ReadyToRunInfo> m_pReadyToRunInfo;
+    IMDInternalImport *m_pManifestMetadata;
+    PEImage *m_pPeImage;
```

Comment:
It looks like the only reason for PEImage is GetNativeMDImport. Sadly it looks like this is why we also need the R2R header detection logic in pedecoder as the manifest metadata decoder depends on the presence of the header. Please let me know what are your thoughts towards disentangling this - I guess most of the craziness stems from the fact that Andon reused the NGen native metadata concept for this purpose.

Reply:
It is trivial to open the metadata from a blob. Is it possible to just do that in NativeImage constructor instead of going round about way through PEImage to do it?

The use of manifest metadata in zapsig decoder is special cases already, so I assume that one should be fine.

Reply:
Fixed in 12th commit.